executor/pipeline/fetch.rs

```diff
@@ -316,27 +466,71 @@ fn execute_attributes_list<'a>(
     thing_manager: Arc<ThingManager>,
 ) -> Result<DocumentList, FetchExecutionError> {
     let mut list = DocumentList::new();
-    let mut iter = object.get_has_type_unordered(snapshot.as_ref(), thing_manager.as_ref(), attribute_type);
+    let mut iter = prepare_attribute_type_has_iterator(object, attribute_type.clone(), &snapshot, &thing_manager)?;
+
     while let Some(result) = iter.next() {
-        match result {
-            Ok((attribute, count)) => {
-                for _ in 0..count {
-                    list.list.push(DocumentNode::Leaf(DocumentLeaf::Concept(Concept::Thing(Thing::Attribute(
-                        attribute.as_reference().into_owned(),
-                    )))));
-                }
+        let (has, count) = result.map_err(|source| FetchExecutionError::ConceptRead { source })?;
+        let attribute = has.attribute();
+        let suitable = attribute
+            .type_()
+            .is_subtype_transitive_of_or_same(snapshot.as_ref(), thing_manager.type_manager(), attribute_type.clone())
+            .map_err(|source| FetchExecutionError::ConceptRead { source })?;
+        if suitable {
+            for _ in 0..count {
+                list.list.push(DocumentNode::Leaf(DocumentLeaf::Concept(Concept::Thing(Thing::Attribute(
+                    has.attribute().into_owned(),
+                )))));
             }
-            Err(err) => return Err(FetchExecutionError::ConceptReadError { source: err }),
         }
     }
     Ok(list)
 }
 
+fn prepare_attribute_type_has_iterator<'a>(
+    object: impl ObjectAPI<'a>,
+    attribute_type: AttributeType<'static>,
+    snapshot: &Arc<impl ReadableSnapshot>,
+    thing_manager: &Arc<ThingManager>,
+) -> Result<HasIterator, FetchExecutionError> {
+    let subtypes = attribute_type
+        .get_subtypes_transitive(snapshot.as_ref(), thing_manager.type_manager())
+        .map_err(|source| FetchExecutionError::ConceptRead { source })?;
+    let attribute_types = TypeAPI::chain_types(attribute_type.clone(), subtypes.into_iter().cloned());
```

Comment:
great thank you for fixing t his!

Reply:
Oh actually this will be incorrect - you need to filter the iterator against the attribute types again! The `Range` literally means range from Min Attributes -> Max attribute! (eg if you say attribute range `(0 to 10)` where you really want `(0, 1, 5, 10)` you'll get `(0, 1, 2, 3...10)`!

Reply:
Dmitrii was meant to fix the API: it should take a `KeyRange<AttributeType>` instead of an Iterator of attribute types! Do you think you could edit it in the ThingManager `get_has_from_thing_to_type_range_unordered` (see `get_has_reverse_by_attribute_and_owner_type_range` for an example of what it should look like)

Reply:
Check the usages of this function. I just prepare the iterator here, and I additionally filter out non-suitable attributes.

Reply:
```
let mut iter = prepare_attribute_type_has_iterator(object, attribute_type.clone(), &snapshot, &thing_manager)?;

    while let Some(result) = iter.next() {
        let (has, count) = result.map_err(|source| FetchExecutionError::ConceptRead { source })?;
        let attribute = has.attribute();
        let suitable = attribute
            .type_()
            .is_subtype_transitive_of_or_same(snapshot.as_ref(), thing_manager.type_manager(), attribute_type.clone())
            .map_err(|source| FetchExecutionError::ConceptRead { source })?;
        if suitable {
        .....
        
        ```

Reply:
Ah ok thanks!